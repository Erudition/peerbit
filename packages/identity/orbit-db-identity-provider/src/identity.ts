
import { Identities } from "./identities";
import { isDefined } from "./is-defined";
import { variant, field, serialize, vec, option } from '@dao-xyz/borsh';
import { createHash } from "crypto";
import { bufferSerializer, U8IntArraySerializer, arraysEqual } from "@dao-xyz/borsh-utils";
import { PublicSignKey } from "@dao-xyz/peerbit-crypto";

@variant(0)
export class Signatures {

  @field(U8IntArraySerializer)
  id: Uint8Array;

  @field({ type: option(U8IntArraySerializer) })
  publicKey?: Uint8Array;

  constructor(options?: {
    id: Uint8Array;
    publicKey?: Uint8Array;
  }) {
    if (options) {
      if (!isDefined(options.id)) {
        throw new Error('Signature of the id (idSignature) is required')
      }

      this.id = options.id;
      this.publicKey = options.publicKey;
    }
  }
  equals(other: Signatures): boolean {
    return arraysEqual(this.id, other.id) && arraysEqual(this.publicKey, other.publicKey)
  }
}

/* const identityToString = (identity: IdentitySerializable | Identity) => {
  return identity.type + "/" + Buffer.from(identity.id).toString('base64')
} */

@variant(0)
export class Identity {

  @field({ type: PublicKey })
  id: PublicKey;

  @field({ type: PublicKey })
  publicKey: PublicKey;

  @field({ type: Signatures })
  signatures: Signatures;

  constructor(
    options?: {
      id: PublicKey,
      publicKey: PublicKey,
      signatures: Signatures
    }
  ) {
    if (options) {

      this.id = options.id;
      this.publicKey = options.publicKey;
      this.signatures = options.signatures;
    }

  }

  public hashCode() {
    return createHash('sha1').update(serialize(this)).digest('hex')
  }

  static from(identity: Identity | any) {
    if (identity instanceof Identity)
      return identity;
    return new Identity({
      id: identity.id,
      publicKey: identity.publicKey,
      signatures: new Signatures({
        id: identity.signatures.id,
        publicKey: identity.signatures.publicKey
      })
    })
  }

  equals(other: Identity): boolean {
    return this.id.equals(other.id) && this.publicKey.equals(other.publicKey) && this.signatures.equals(other.signatures)
  }

  /* toString() {
    return identityToString(this);
  } */
}


/* 
export class Identity {

  _id: Uint8Array;
  _publicKey: Ed25519PublicKey;
  _signatures: Signatures;
  _type: string;


  _provider: Identities;
  constructor(options?: { id: Uint8Array, publicKey: Ed25519PublicKey, signatures: Signatures, type: string, provider: Identities }) {
    if (options) {
      if (!isDefined(options.id)) {
        throw new Error('Identity id is required')
      }

      if (!isDefined(options.publicKey)) {
        throw new Error('Invalid public key')
      }

      if (!isDefined(options.signatures)) {
        throw new Error('Signatures are required')
      }

      if (!isDefined(options.type)) {
        throw new Error('Identity type is required')
      }

      if (!isDefined(options.provider)) {
        throw new Error('Identity provider is required')
      }

      this._id = options.id
      this._publicKey = options.publicKey
      this._signatures = options.signatures
      this._type = options.type
      this._provider = options.provider
    }
  }


  get id(): Uint8Array {
    return this._id
  }

  get publicKey(): Ed25519PublicKey {
    return this._publicKey
  }

  get signatures(): Signatures {
    return this._signatures
  }

  get type(): string {
    return this._type
  }

  get provider(): Identities {
    return this._provider
  }

  toSerializable(): IdentitySerializable {
    const ser = new IdentitySerializable();
    ser.id = this.id;
    ser.publicKey = this.publicKey;
    ser.signatures = new Signatures({ ...this.signatures });
    return ser;
  }

  static isIdentity(identity) {
    return identity.id !== undefined &&
      identity.publicKey !== undefined &&
      identity.signatures !== undefined &&
      identity.signatures.id !== undefined &&
      identity.signatures.publicKey !== undefined &&
      identity.type !== undefined
  }

  equals(other: IdentitySerializable): boolean {
    return arraysEqual(this.id, other.id) && Buffer.compare(this.publicKey.getBuffer(), other.publicKey.getBuffer()) === 0 && this.type === other.type && this.signatures.equals(other.signatures)
  }

} */




/*  toString() {
   return identityToString(this);
 } */


/* static from(identity: IdentitySerializable): Identity {
  return new Identity({
    id: identity.id,
    publicKey: identity.publicKey,
    signatures: new Signatures({
      id: identity.signatures.id,
      publicKey: identity.signatures.publicKey
    }),
    type: identity.type
  })
} */