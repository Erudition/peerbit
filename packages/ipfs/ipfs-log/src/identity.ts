import {
  Ed25519PrivateKey,
  Ed25519PublicKey,
  Secp256k1PublicKey,
} from "@dao-xyz/peerbit-crypto";

/**
 * Can sign
 */
export type Secp256k1Identity = {
  publicKey: Secp256k1PublicKey;
  sign: (data: Uint8Array) => Promise<Uint8Array>;
};

/**
 * Can sign and send/recieve encrypted messages
 */
export type Ed25519Identity = {
  publicKey: Ed25519PublicKey;
  privateKey: Ed25519PrivateKey;
  sign: (data: Uint8Array) => Promise<Uint8Array>;
};

export type Identity = Ed25519Identity | Secp256k1Identity;
