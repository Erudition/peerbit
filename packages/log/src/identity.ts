import {
    Ed25519PrivateKey,
    Ed25519PublicKey,
    Secp256k1Keccak256PublicKey,
} from "@dao-xyz/peerbit-crypto";

/**
 * Can sign
 */
export type Secp256k1Identity = {
    publicKey: Secp256k1Keccak256PublicKey;
    sign: (data: Uint8Array) => Promise<Uint8Array> | Uint8Array;
};

/**
 * Can sign and send/recieve encrypted messages
 */
export type Ed25519Identity = {
    publicKey: Ed25519PublicKey;
    privateKey: Ed25519PrivateKey;
    sign: (data: Uint8Array) => Promise<Uint8Array> | Uint8Array;
};

export type Identity = Ed25519Identity | Secp256k1Identity;
