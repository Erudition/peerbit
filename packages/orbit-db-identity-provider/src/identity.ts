
import { Identities } from "./identities";
import { isDefined } from "./is-defined";
import { variant, field, serialize } from '@dao-xyz/borsh';
import { createHash } from "crypto";
export type IdentityProviderType = 'orbitdb' | 'ethereum' | 'solana' | string;


@variant(0)
export class Signatures {

  id: string;
  publicKey: string;

  constructor(options?: {
    id: string;
    publicKey: string;
  }) {
    if (options) {
      if (!isDefined(options.id)) {
        throw new Error('Signature of the id (idSignature) is required')
      }

      if (!isDefined(options.publicKey)) {
        throw new Error('Signature of (publicKey + idSignature) is required')
      }
      this.id = options.id;
      this.publicKey = options.publicKey;
    }
  }

}

@variant(0)
export class IdentitySerializable {

  @field({ type: 'String' })
  id: string;

  @field({ type: 'String' })
  publicKey: string;

  @field({ type: Signatures })
  signatures: Signatures;

  @field({ type: 'String' })
  type: IdentityProviderType;
  constructor(
    options?: {
      id: string,
      publicKey: string,
      signatures: Signatures,
      type: string
    }
  ) {
    if (options) {

      this.id = options.id;
      this.publicKey = options.publicKey;
      this.signatures = options.signatures;
      this.type = options.type;
    }

  }

  public hashCode() {
    return createHash('sha1').update(serialize(this)).digest('hex')
  }

  static from(identity: IdentitySerializable | any) {
    if (identity instanceof IdentitySerializable)
      return identity;
    return new IdentitySerializable({
      id: identity.id,
      publicKey: identity.publicKey,
      signatures: new Signatures({
        id: identity.signatures.id,
        publicKey: identity.signatures.publicKey
      }),
      type: identity.type
    })
  }
}



export class Identity {

  _id: string;
  _publicKey: string;
  _signatures: Signatures;
  _type: string;


  _provider: Identities;
  constructor(options?: { id: string, publicKey: string, signatures: Signatures, type: string, provider: Identities }) {
    if (options) {
      if (!isDefined(options.id)) {
        throw new Error('Identity id is required')
      }

      if (!isDefined(options.publicKey)) {
        throw new Error('Invalid public key')
      }

      if (!isDefined(options.signatures)) {
        throw new Error('Signatures are required')
      }

      if (!isDefined(options.type)) {
        throw new Error('Identity type is required')
      }

      if (!isDefined(options.provider)) {
        throw new Error('Identity provider is required')
      }

      this._id = options.id
      this._publicKey = options.publicKey
      this._signatures = options.signatures
      this._type = options.type
      this._provider = options.provider
    }
  }

  /**
  * This is only used as a fallback to the clock id when necessary
  * @return {string} public key hex encoded
  */
  get id(): string {
    return this._id
  }

  get publicKey(): string {
    return this._publicKey
  }

  get signatures(): Signatures {
    return this._signatures
  }

  get type(): string {
    return this._type
  }

  get provider(): Identities {
    return this._provider
  }

  toSerializable(): IdentitySerializable {
    const ser = new IdentitySerializable();
    ser.id = this.id;
    ser.publicKey = this.publicKey;
    ser.signatures = new Signatures({ ...this.signatures });
    ser.type = this.type;
    return ser;
  }

  static isIdentity(identity) {
    return identity.id !== undefined &&
      identity.publicKey !== undefined &&
      identity.signatures !== undefined &&
      identity.signatures.id !== undefined &&
      identity.signatures.publicKey !== undefined &&
      identity.type !== undefined
  }


  /* static from(identity: IdentitySerializable): Identity {
    return new Identity({
      id: identity.id,
      publicKey: identity.publicKey,
      signatures: new Signatures({
        id: identity.signatures.id,
        publicKey: identity.signatures.publicKey
      }),
      type: identity.type
    })
  } */
}



