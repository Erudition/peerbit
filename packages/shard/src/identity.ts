import IdentityProvider from "orbit-db-identity-provider/src/identity-provider-interface";

import { Keypair, PublicKey as SPublicKey } from '@solana/web3.js';
import { Identity } from "@dao-xyz/orbit-db-identity-provider";
import nacl from "tweetnacl";
import bs58 from 'bs58';



export class SolanaIdentityProvider extends IdentityProvider {

    wallet: SPublicKey;
    keypair: Keypair | undefined;
    signer: ((data: Uint8Array) => any) | undefined
    constructor(options: { wallet: SPublicKey, keypair?: Keypair, signer?: (data: Uint8Array) => any }) {
        super()
        this.wallet = options.wallet;
        this.keypair = options.keypair;
        this.signer = options.signer;
    }

    // Returns the type of the identity provider
    static get type(): string {
        return 'solana'
    }

    // Returns the signer's id
    async getId(): Promise<string> {
        return this.wallet.toBase58();
    }

    // Returns a signature of pubkeysignature
    async signIdentity(data: string) {
        if (!this.signer && !this.keypair) {
            throw new Error("This identity to not support signing")
        }

        let array = Uint8Array.from(Buffer.from(data));
        return this.signer ? this.signer(array) : nacl.sign(array, this.keypair.secretKey)
    }

    static async verifyIdentity(identity: Identity) {

        // Verify that identity was signed by the id
        /* const signedKey = nacl.sign.open(Uint8Array.from(Buffer.from(identity.signatures.publicKey)), bs58.decode(identity.id));
        let verified = nacl.verify(signedKey, Uint8Array.from(Buffer.from(identity.publicKey + identity.signatures.id)));
        return verified */
        return false;
    }
}
