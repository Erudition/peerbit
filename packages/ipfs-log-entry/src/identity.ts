import { LamportClock as Clock, LamportClock } from './lamport-clock'
import { isDefined } from './is-defined'
import { variant, field, vec, option, serialize, deserialize, BinaryReader } from '@dao-xyz/borsh';
import io from '@dao-xyz/orbit-db-io';
import { IPFS } from 'ipfs-core-types/src/'
import { Identities, Identity, IdentitySerializable } from '@dao-xyz/orbit-db-identity-provider'
import { arraysEqual, U8IntArraySerializer, U8IntArraySerializerOptional } from '@dao-xyz/io-utils';
import { X25519PublicKey, Ed25519PublicKey } from 'sodium-plus';
import { CryptOptions, DecryptedThing, EncryptedThing, MaybeEncrypted } from './encryption';
import { IOOptions } from './entry';
/* 
@variant(0)
export class IdentitySecure {

    @field({ type: MaybeEncrypted })
    _identity: MaybeEncrypted<IdentitySerializable>;

    constructor(props?: {
        identity: MaybeEncrypted<IdentitySerializable>
    }) {
        if (props) {
            this._identity = props.identity;
        }
    }

    init(crypt?: CryptOptions) {
        this._identity.init(crypt);
        return this;
    }
    get identity(): Promise<MaybeEncrypted<IdentitySerializable>> {
        return this._identity.decrypt()
    }
}
 */
@variant(0)
export class IdentityWithSignature {

    @field({ type: IdentitySerializable })
    _identity: IdentitySerializable

    @field(U8IntArraySerializerOptional)
    _signature: Uint8Array; // Signing some data

    constructor(props: {
        identity: IdentitySerializable,
        signature: Uint8Array;
    }) {
        if (props) {
            this._identity = props.identity;
            this._signature = props.signature;
        }
    }

    equals(other: IdentityWithSignature): boolean {
        return this._identity.equals(other._identity) && arraysEqual(this._signature, other._signature)
    }
}

@variant(0)
export class IdentityWithSignatureSecure {

    @field({ type: MaybeEncrypted })
    _identityWithSignature: MaybeEncrypted<IdentityWithSignature>;

    constructor(props?: {
        identityWithSignature: MaybeEncrypted<IdentityWithSignature> | IdentityWithSignature
    }) {
        if (props) {
            if (props.identityWithSignature instanceof IdentityWithSignature) {
                this._identityWithSignature = new DecryptedThing({
                    data: serialize(props.identityWithSignature)
                })
            }
            else {
                this._identityWithSignature = props.identityWithSignature;
            }
        }
    }

    init(crypt?: CryptOptions) {
        this._identityWithSignature.init(crypt);
        return this;
    }

    get identity(): Promise<IdentitySerializable> {
        return this._identityWithSignature.decrypt().then(x => x.getValue(IdentityWithSignature)._identity)
    }

    get signature(): Promise<Uint8Array> {
        return this._identityWithSignature.decrypt().then(x => x.getValue(IdentityWithSignature)._signature)
    }

    async encrypt(recieverPublicKey: X25519PublicKey) {
        if (this._identityWithSignature instanceof EncryptedThing) {
            return;
        }
        else if (this._identityWithSignature instanceof DecryptedThing) {
            this._identityWithSignature = await this._identityWithSignature.encrypt(recieverPublicKey)

        }
        else {
            throw new Error("Unsupported")
        }
    }
    equals(other: IdentityWithSignatureSecure) {
        return this._identityWithSignature.equals(other._identityWithSignature);
    }
}