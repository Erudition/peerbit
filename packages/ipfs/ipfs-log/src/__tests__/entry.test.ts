import assert from 'assert';
import rmrf from 'rimraf'
import fs from 'fs-extra'
import { Entry, Payload } from '../entry.js';
import { createStore, Keystore, KeyWithMeta } from '@dao-xyz/orbit-db-keystore'
import { deserialize, serialize } from '@dao-xyz/borsh';
import { Ed25519Keypair, Ed25519PublicKey, X25519Keypair, X25519PublicKey } from '@dao-xyz/peerbit-crypto';
import sodium from 'libsodium-wrappers';
// Test utils
import { jest } from '@jest/globals';
import {
  nodeConfig as config,
  Session
} from '@dao-xyz/orbit-db-test-utils'
import { IPFS } from 'ipfs-core-types';
import { dirname } from 'path';
import { fileURLToPath } from 'url';
import path from 'path';
import { Identity } from '../identity.js';

const __filename = fileURLToPath(import.meta.url);
const __filenameBase = path.parse(__filename).base;
const __dirname = dirname(__filename);

const identityFromSignKey = (key: KeyWithMeta<Ed25519Keypair>): Identity => {
  if (!key) {
    throw new Error("Key not defined");
  }
  return {
    ...key.keypair,
    sign: async (data: Uint8Array) => (await key.keypair.sign(data))
  }
}
const API = 'js-ipfs';

describe('Entry', function () {
  jest.setTimeout(config.timeout)
  let session: Session, ipfs: IPFS

  const { signingKeyFixtures, signingKeysPath } = config
  let keystore: Keystore, signKey: KeyWithMeta<Ed25519Keypair>

  beforeAll(async () => {
    await sodium.ready;
    session = await Session.connected(1, API, config.defaultIpfsConfig);
    ipfs = session.peers[0].ipfs;
    await sodium.ready;


    await fs.copy(signingKeyFixtures(__dirname), signingKeysPath(__filenameBase))

    keystore = new Keystore(await createStore(signingKeysPath(__filenameBase)));
    await keystore.waitForOpen();

    signKey = await keystore.getKey(new Uint8Array([0])) as KeyWithMeta<Ed25519Keypair>
  })

  afterAll(async () => {
    await session.stop();

    await fs.copy(signingKeyFixtures(__dirname), signingKeysPath(__filenameBase))

    rmrf.sync(signingKeysPath(__filenameBase))
    await keystore?.close()

  })
  describe('endocing', () => {
    it('can serialize and deserialialize', async () => {

      const entry = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello'
      })
      deserialize(serialize(entry), Entry)
    })
  })

  describe('create', () => {
    it('creates a an empty entry', async () => {
      const expectedHash = 'zdpuAodEgpKQpFjAgCtJPYLDFjLrYJ56j69MvhS7gdrvEVisW'
      const entry = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello'
      })

      expect(entry.hash).toEqual(expectedHash)
      expect(entry.gid).toEqual(await sodium.crypto_generichash(32, 'A'))
      assert.deepStrictEqual(entry.clock.id, new Ed25519PublicKey({
        publicKey: signKey.keypair.publicKey.publicKey
      }).bytes);
      expect(entry.clock.time).toEqual(0n)
      expect(entry.payload.getValue()).toEqual('hello')
      expect(entry.next.length).toEqual(0)
    })

    it('creates a entry with payload', async () => {
      const expectedHash = 'zdpuAotgKm27uJe3j1dJTHi29P9Kunt1oKNDdcutcjY4gWY8k'
      const payload = 'hello world'
      const entry = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload, next: []
      })
      expect(entry.hash).toEqual(expectedHash)
      expect(entry.payload.getValue()).toEqual(payload)
      expect(entry.gid).toEqual(await sodium.crypto_generichash(32, 'A'))
      assert.deepStrictEqual(entry.clock.id, new Ed25519PublicKey({
        publicKey: signKey.keypair.publicKey.publicKey
      }).bytes);
      expect(entry.clock.time).toEqual(0n)
      expect(entry.next.length).toEqual(0)
    })

    it('creates a encrypted entry with payload', async () => {

      const payload = 'hello world'
      const senderKey = await keystore.createX25519Key({ id: 'sender', overwrite: true });
      const receiverKey = await keystore.createX25519Key({ id: 'reciever', overwrite: true });
      const entry = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload, next: [], encryption: {
          reciever: {
            clock: undefined,
            signature: undefined,
            payload: receiverKey.keypair.publicKey
          },
          options: {
            getEncryptionKeypair: () => Promise.resolve(senderKey.keypair as Ed25519Keypair | X25519Keypair),
            getAnyKeypair: async (publicKeys: X25519PublicKey[]) => {
              for (let i = 0; i < publicKeys.length; i++) {
                if (publicKeys[i].equals((senderKey.keypair as X25519Keypair).publicKey)) {
                  return {
                    index: i,
                    keypair: senderKey.keypair as Ed25519Keypair | X25519Keypair
                  }
                }
                if (publicKeys[i].equals((receiverKey.keypair as X25519Keypair).publicKey)) {
                  return {
                    index: i,
                    keypair: receiverKey.keypair as Ed25519Keypair | X25519Keypair
                  }
                }

              }
            }
          }
        }
      })
      assert(entry.payload instanceof Payload)
      expect(entry.payload.getValue()).toEqual(payload);

      // We can not have a hash check because nonce of encryption will always change
      expect(entry.gid).toEqual(await sodium.crypto_generichash(32, 'A'))
      assert.deepStrictEqual(entry.clock.id, (new Ed25519PublicKey({ publicKey: signKey.keypair.publicKey.publicKey })).bytes)
      expect(entry.clock.time).toEqual(0n)
      expect(entry.next.length).toEqual(0)
    })

    it('creates a entry with payload and next', async () => {
      const expectedHash = 'zdpuAvEB1kk7M6ZhrhV5eG85pKEiArEbHV2EFgA37qcuk5w1p'
      const payload1 = 'hello world'
      const payload2 = 'hello again'
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload1, next: []
      })
      const clock = entry1.clock;
      /* entry1.metadata = new MetadataSecure({
        metadata: new Metadata({
          id: await entry1.metadata.id,
          identity: await entry1.metadata.identity,
          signature: await entry1.metadata.signature
        })
      }) */
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload2, next: [entry1]
      })
      expect(entry2.payload.getValue()).toEqual(payload2)
      expect(entry2.next.length).toEqual(1)
      expect(entry2.maxChainLength).toEqual(2n); // because 1 next
      expect(entry2.hash).toEqual(expectedHash)
      assert.deepStrictEqual(entry2.clock.id, new Ed25519PublicKey({
        publicKey: signKey.keypair.publicKey.publicKey
      }).bytes);
      expect(entry2.clock.time).toEqual(1n)
    })

    it('`next` parameter can be an array of strings', async () => {
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello1', next: []
      })
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello2', next: [entry1]
      })
      assert.strictEqual(typeof entry2.next[0] === 'string', true)
    })

    it('`next` parameter can be an array of Entry instances', async () => {
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello1', next: []
      })
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello2', next: [entry1]
      })
      assert.strictEqual(typeof entry2.next[0] === 'string', true)
    })

    it('can calculate join gid from `next` max chain length', async () => {
      const entry0A = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello1', next: []
      })

      const entry1A = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello1', next: [entry0A]
      })

      const entry1B = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'B', clock: entry1A.clock, data: 'hello1', next: []
      })

      expect(entry1A.gid > entry1B.gid); // so that gid is not choosen because A has smaller gid
      expect(entry1A.clock.time).toEqual(entry1B.clock.time);

      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'Should not be used', data: 'hello2', next: [entry1A, entry1B]
      })
      expect(entry2.gid).toEqual(entry1A.gid) // because A has alonger chain
    })

    it('can calculate join gid from `next` max clock', async () => {

      const entry1A = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'B', data: 'hello1', next: []
      })

      const entry1B = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', clock: entry1A.clock.advance(), data: 'hello1', next: []
      })

      expect(entry1B.gid > entry1A.gid); // so that gid is not choosen because B has smaller gid
      expect(entry1A.maxChainLength).toEqual(entry1B.maxChainLength);
      expect(entry1A.clock.time).toEqual(entry1B.clock.time - 1n);

      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'Should not be used', data: 'hello2', next: [entry1A, entry1B]
      })
      expect(entry2.gid).toEqual(entry1B.gid) // because A has alonger chain
    })

    it('can calculate join gid from `next` gid comparison', async () => {

      const entry1A = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'B', data: 'hello1', next: []
      })

      const entry1B = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', clock: entry1A.clock, data: 'hello1', next: []
      })

      expect(entry1B.gid < entry1A.gid); // so that B is choosen because of gid
      expect(entry1A.maxChainLength).toEqual(entry1B.maxChainLength);
      expect(entry1A.clock.time).toEqual(entry1B.clock.time);

      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'Should not be used', data: 'hello2', next: [entry1A, entry1B]
      })
      expect(entry2.gid).toEqual(entry1B.gid) // because gid B < gid A
    })

    it('can calculate reuse gid from `next`', async () => {
      const entry1A = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello1', next: []
      })

      const entry1B = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gid: entry1A.gid, data: 'hello1', next: []
      })

      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'Should not be used', data: 'hello2', next: [entry1A, entry1B]
      })
      expect(entry2.gid).toEqual(entry1A.gid)
      expect(entry1A.gid).toEqual(entry1B.gid)
    })


    it('will use next for gid instaed of gidSeed', async () => {
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello1', next: []
      })


      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'Should not be used', data: 'hello2', next: [entry1]
      })
      expect(entry2.gid).toEqual(entry1.gid)

    })


    it('throws an error if data is not defined', async () => {
      let err: any
      try {
        await Entry.create({
          ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: null, next: []
        })
      } catch (e: any) {
        err = e
      }
      expect(err.message).toEqual('Entry requires data')
    })

    it('throws an error if next is not an array', async () => {
      let err: any
      try {
        await Entry.create({
          ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello', next: {} as any
        })
      } catch (e: any) {
        err = e
      }
      expect(err.message).toEqual('\'next\' argument is not an array')
    })
  })

  describe('toMultihash', () => {

    it('returns an ipfs multihash', async () => {
      const expectedMultihash = 'zdpuAodEgpKQpFjAgCtJPYLDFjLrYJ56j69MvhS7gdrvEVisW'
      const entry = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello', next: []
      })
      const multihash = await Entry.toMultihash(ipfs, entry)
      expect(multihash).toEqual(expectedMultihash)
    })


    /*  TODO what is the point of this test?
    
    it('throws an error if the object being passed is invalid', async () => {
      let err
      try {
        const entry = await Entry.create({ ipfs, identity: identityFromSignKey(signKey), gidSeed:   'A', data: 'hello', next: [] })
        delete ((entry.metadata as MetadataSecure)._metadata as DecryptedThing<Metadata>)
        await Entry.toMultihash(ipfs, entry)
      } catch (e: any) {
        err = e
      }
      expect(err.message).toEqual('Invalid object format, cannot generate entry hash')
    }) */
  })

  describe('fromMultihash', () => {
    it('creates a entry from ipfs hash', async () => {
      const expectedHash = 'zdpuAvEB1kk7M6ZhrhV5eG85pKEiArEbHV2EFgA37qcuk5w1p'
      const payload1 = 'hello world'
      const payload2 = 'hello again'
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload1, next: []
      })
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload2, next: [entry1]
      })
      const final = await Entry.fromMultihash<string>(ipfs, entry2.hash)
      final.init(entry2);
      assert(final.equals(entry2));
      expect(final.gid).toEqual((await sodium.crypto_generichash(32, 'A')))
      expect(final.payload.getValue()).toEqual(payload2)
      expect(final.next.length).toEqual(1)
      expect(final.next[0]).toEqual(entry1.hash)
      expect(final.hash).toEqual(expectedHash)
    })
  })

  describe('isParent', () => {
    it('returns true if entry has a child', async () => {
      const payload1 = 'hello world'
      const payload2 = 'hello again'
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload1, next: []
      })
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload2, next: [entry1]
      })
      expect(Entry.isDirectParent(entry1, entry2)).toEqual(true);
    })

    it('returns false if entry does not have a child', async () => {
      const payload1 = 'hello world'
      const payload2 = 'hello again'
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload1, next: []
      })
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload2, next: []
      })
      const entry3 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload2, next: [entry2]
      })
      expect(Entry.isDirectParent(entry1, entry2)).toEqual(false);
      expect(Entry.isDirectParent(entry1, entry1)).toEqual(false);
      expect(Entry.isDirectParent(entry2, entry3)).toEqual(true);

    })
  })

  describe('compare', () => {
    it('returns true if entries are the same', async () => {
      const payload1 = 'hello world'
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload1, next: []
      })
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload1, next: []
      })
      expect(Entry.isEqual(entry1, entry2)).toEqual(true);
    })

    it('returns true if entries are not the same', async () => {
      const payload1 = 'hello world1'
      const payload2 = 'hello world2'
      const entry1 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload1, next: []
      })
      const entry2 = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: payload2, next: []
      })
      expect(Entry.isEqual(entry1, entry2)).toEqual(false);
    })
  })

  describe('isEntry', () => {
    it('is an Entry', async () => {
      const entry = await Entry.create({
        ipfs, identity: identityFromSignKey(signKey), gidSeed: 'A', data: 'hello', next: []
      })
      expect(Entry.isEntry(entry)).toEqual(true)
    })

    it('is not an Entry - no id', async () => {
      const fakeEntry = { data: { v: 1, hash: 'Foo', payload: 123, seq: 0 }, next: [], }
      expect(Entry.isEntry(fakeEntry as any)).toEqual(false)
    })

    it('is not an Entry - no seq', async () => {
      const fakeEntry = { data: { v: 1, hash: 'Foo', payload: 123 }, next: [] }
      expect(Entry.isEntry(fakeEntry as any)).toEqual(false)
    })

    it('is not an Entry - no next', async () => {
      const fakeEntry = { data: { id: 'A', v: 1, hash: 'Foo', seq: 0 }, payload: 123 }
      expect(Entry.isEntry(fakeEntry as any)).toEqual(false)
    })

    it('is not an Entry - no version', async () => {
      const fakeEntry = { data: { id: 'A', payload: 123, seq: 0 }, next: [] }
      expect(Entry.isEntry(fakeEntry as any)).toEqual(false)
    })

    it('is not an Entry - no hash', async () => {
      const fakeEntry = { data: { id: 'A', v: 1, payload: 123, seq: 0 }, next: [] }
      expect(Entry.isEntry(fakeEntry as any)).toEqual(false)
    })

    it('is not an Entry - no payload', async () => {
      const fakeEntry = { data: { id: 'A', v: 1, hash: 'Foo', seq: 0 }, next: [] }
      expect(Entry.isEntry(fakeEntry as any)).toEqual(false)
    })
  })
})