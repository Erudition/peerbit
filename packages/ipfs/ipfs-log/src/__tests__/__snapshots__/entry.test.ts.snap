// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry create creates a an empty entry 1`] = `"zb2rhoAE1yKQhNxuUytaMN2D1kMGaYXVQCmpu6PiZfmsFo6PR"`;

exports[`Entry create creates a entry with payload 1`] = `"zb2rheoYTrZCX5PiJbqAFYT9fSNLGDR1Y9oEn42uSRV5U7L5A"`;

exports[`Entry create creates a entry with payload and next 1`] = `"zb2rhXHEt9FdqQ81EcQ2WXnQ6BMqRXXvz98umwxiVZw995vXT"`;

exports[`Entry fromMultihash creates a entry from ipfs hash 1`] = `"zb2rhnoM32vP8ZXhvTVHhVCEKBMwqwYEuUiPKhzZZA2t8qsfU"`;

exports[`Entry toMultihash returns an ipfs multihash 1`] = `"zb2rhoAE1yKQhNxuUytaMN2D1kMGaYXVQCmpu6PiZfmsFo6PR"`;
